### PR TITLE
20160623

### LINKED ISSUE #680: check aws credentials have enough permissions before install

### LINKED ISSUE #720: Security Group inbound rules
A couple of the security groups created by Convox have inbound rules allowing 0.0.0.0/0 where I would have expected to see my VPC CIDR. For example, I have a linked process in my `docker-compose.yml` with a single port, correctly resulting in the creation of an internal load balancer. However, the associated security group inbound rules look as follows:

internal elb inbound

Looking at the [AWS documentation](http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/elb-security-groups.html#elb-vpc-security-groups), I would have expected to see my VPC CIDR instead of 0.0.0.0/0. The security group with tag value "BalancerSecurityGroup" also has some 0.0.0.0/0 rules. The security groups for the web process, redis service, postgres service, and the one with tag value "SecurityGroup" all look as I would have expected.

Happy to provide additional info if needed.

### LINKED ISSUE #781: Build failing for docker 1.12 due to `docker tag -f` flag which is now removed
We've done it folks. Deprecation warning since 1.10 and have we done? Ignore it!

Fix should be just to remove `-f` altogether from all references of `docker tag` such as the offending https://github.com/convox/rack/blob/c533520b5e587e6e44253fb0240bab618c355bc4/api/manifest/manifest.go#L349 that prevents `convox start` from working with this error:

```
Successfully built 52b83b62a098
RUNNING: docker tag -f convox-hkrfeetpkt web-dev/web
unknown shorthand flag: 'f' in -f
See 'docker tag --help'.
ERROR: exit status 1
```

### LINKED ISSUE #794: convox ps formatting error
```
ID            NAME      RELEASE      CPU             SIZE  STARTED     COMMAND
07ebeea7a189  postgres  RIDZBQFKOWC  %!d(float64=0)  256   2 days ago  /docker-entrypoint.sh postgres
12dce2ff5e16  redis     RIDZBQFKOWC  %!d(float64=0)  256   2 days ago  /docker-entrypoint.sh redis-server /tmp/redis.conf
```

### PR DESCRIPTION
**Pull request by MiguelMoll**
Fixes #720 Security Group inbound rules
Fixes #680 Check aws credentials have enough permissions before install
Fixes #794 PS output formatting
Fixes #781 (By upgrading to docker 1.11)

## Multi-PR Release Playbook
- [x] Create a pre-release branch
- [x] Merge other PRs to include in the release
- [x] Release branch (20160623135412-20160623)
- [x] Pass CI
- [ ] Code review
- [ ] Create release notes and docs PRs
- [ ] Merge into master
- [ ] Release master
- [ ] Pass CI
- [ ] Update dev and testing racks
- [ ] If rack or app formation changes, test upgrade → re-deploy existing app → downgrade cycle
- [ ] Create [Rack release record](https://github.com/convox/rack/releases) and/or update [docs](https://github.com/convox/convox.github.io)
- [ ] Publish release
- [ ] Release CLI

### CODE REVIEW

File: api/models/app.go [hunk 1]
```diff
@@ -645,25 +636,23 @@ func (a *App) RunAttached(process, command, releaseId string, height, width int,
...
 		return err
 	}
 
 	err = d.ResizeContainerTTY(res.ID, height, width)
-
 	if err != nil {
-		return err
+		// In some cases, a container might finish and exit by the time ResizeContainerTTY is called.
+		// Resizing the TTY shouldn't cause the call to error out for cases like that.
+		fmt.Printf("fn=RunAttached level=warning msg=\"unable to resize container: %s\"", err)
```
Comment (nzoschke) [inline api/models/app.go]: I like this. Starting to see a pattern with deciding that some Docker API calls aren't important enough to blow up our API call.

File: api/dist/kernel.json [hunk 2]
```diff
@@ -15,14 +15,14 @@
...
     "RegionConfig": {
-      "us-east-1": { "Ami": "ami-67a3a90d" },
-      "us-west-1": { "Ami": "ami-b7d5a8d7" },
-      "us-west-2": { "Ami": "ami-c7a451a7" },
-      "eu-west-1": { "Ami": "ami-9c9819ef" },
-      "eu-central-1": { "Ami": "ami-9aeb0af5" },
-      "ap-northeast-1": { "Ami": "ami-7e4a5b10" },
-      "ap-southeast-1": { "Ami": "ami-be63a9dd" },
-      "ap-southeast-2": { "Ami": "ami-b8cbe8db" }
+      "us-east-1": { "Ami": "ami-8f7687e2" },
```
Comment (nzoschke) [inline api/dist/kernel.json]: These look good. It would be useful to have an AMI checklist

File: api/models/templates/app.tmpl [hunk 3]
```diff
@@ -332,7 +332,11 @@
         "SecurityGroupIngress": [
           {{ range .PortMappings }}
             {
+              {{ if .Public }}
               "CidrIp": "0.0.0.0/0",
+              {{ else }}
+              "CidrIp": "10.0.0.0/16",
```
Comment (nzoschke) [inline api/models/templates/app.tmpl]: hard coding this is incorrect. We can get VPCCIDR from the rack environment, which will reflect whatever parameter value was used when installing convox.

Comment (nzoschke) [inline api/models/templates/app.tmpl]: I see two strategies here.

1. Use a go template variable and pass `os.Getenv("VPCCIDR")` into the go template rendering context.
2. Add VPCCIDR as an app stack cloudformation parameter, and use a cloudformation `{ "Ref": "VPCCIDR" }'

File: api/models/fixtures/balancer_labels.json [hunk 4]
```diff
@@ -280,6 +280,11 @@
       "Description": "VPC for this app",
       "Type": "AWS::EC2::VPC::Id"
     },
+    "VPCCIDR": {
+      "Default": "",
+      "Description": "VPC CIRDR for this app",
```
Comment (nzoschke) [inline api/models/fixtures/balancer_labels.json]: Typo in "CIRDR"

**Comment by nzoschke** [comment]
Looks good to me!

**Comment by MiguelMoll** [comment]
## Upgrade to docker 1.11

With Docker finally removing the `-f` option (thanks to @asyazwan for bringing it to our attention with #781), we decided the best route is to upgrade to [Docker 1.11](https://github.com/docker/docker/blob/master/CHANGELOG.md#1111-2016-04-26).
This includes the the alpine imaged used by rack itself but also an [AMI optimize for ECS Instances](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html):

```
The current Amazon ECS-optimized AMI (amzn-ami-2016.03.c-amazon-ecs-optimized) consists of:

The latest minimal version of the Amazon Linux AMI
The latest version of the Amazon ECS container agent (1.10.0)
The recommended version of Docker for the latest Amazon ECS container agent (1.11.1)
The latest version of the ecs-init package to run and monitor the Amazon ECS agent (1.10.0-1)
```

See PR #797 for more information.

## Process CPU Configuration

Thanks to @beedub you can now configure CPU shares for any process type.

The default is still the same -- not explicitly changing this setting means that all processes share all of the CPU available on an instance. This is still the sensible default to get the most out of CPUs in the cluster.

But the ability to specify this opens up some more powerful use cases.

If you set a worker CPU shares to a small number with `convox apps params set WorkerCpu=50` it will restrict the amount of CPU -- most importantly the number of CPU cores -- that a single worker can consume. For some apps this is a very good way to assure that the web processes don't get adversely impacted by workers.

See #741 for lots of good discussion about the use cases and implementation.

## Custom App Load Balancers Security Groups

Convox handles provisioning and configuring a load balancer based on the `docker-compose.yml` port mappings:

```yaml
web:
  image: httpd
  ports:
    - 80:80
```

By default this load balancer will be exposed to the entire internet. Sometimes we want to tighten this to only allow connections from a single computer or single network.

You can now apply additional restrictions to this load balancer by assigning a custom security group to an app:

```bash
# get the VPC ID
$ convox rack
Name     convox
Status   running
Version  20160615203231
Region   us-east-1
Count    3
Type     t2.large

$ convox ps --app convox
ID            NAME      RELEASE                  SIZE  STARTED     COMMAND                           
cde9ac958f36  monitor   20160615203231-20160615  64    4 days ago  /go/bin/init sh -c api/bin/monitor
1845eb6ec788  registry  20160615203231-20160615  128   1 day ago   /docker-entrypoint.sh             
79e05a82d06d  registry  20160615203231-20160615  128   4 days ago  /docker-entrypoint.sh             
541d9b8509cb  web       20160615203231-20160615  128   1 day ago   /go/bin/init sh -c api/bin/web    
e9fd0ed66d41  web       20160615203231-20160615  128   4 days ago  /go/bin/init sh -c api/bin/web 

$ convox exec 541d9b8509cb env --app convox | grep VPC
VPC=vpc-f8006b9c

# create a new security group
$ aws ec2 create-security-group --group-name=office --description="office network only" --vpc-id=vpc-f8006b9c
{
    "GroupId": "sg-3bdaaf40"
}

# restrict ingress to a single subnet
$ aws ec2 authorize-security-group-ingress --group-id=sg-3bdaaf40 --protocol=tcp --port=0-65535 --cidr=24.75.239.0/32

# apply the security group to the app load balancers
$ convox apps params set SecurityGroup=sg-3bdaaf40
Updating parameters... OK
```

After this, the service will only be accessable to IPs from the specified subnet, blocking access from the internet at large.

See @MiguelMoll's patch on #775 for more information.

## Build Cleanup

AWS ECR has a default limit of 500 Docker Images, which you can eventually hit after many builds. Thanks to @awsmsrc on #773, convox now automatically deletes the very oldest builds after a successful new build. This both avoids hitting the service limit and saves money on storage costs.

As a reminder, Convox still retains the last 50 builds and releases to facilitate a rollback from a bad release:

```bash
$ convox builds
ID           STATUS    RELEASE      STARTED      ELAPSED  DESC
BOFEDWPIEEB  complete  RLBTJENIMQI  3 days ago   6s           
BVFZQPTJJNV  complete  RTAYZHUSGCJ  1 week ago   22s          
BKYKUURZKQU  complete  RONCLAIMWTP  2 weeks ago  49s          
BFUOFQOMAFI  complete  RDJQJEUAPTP  2 weeks ago  10s          
BHGAKQOJCCA  complete  RQOHWZJYDGG  2 weeks ago  5s       

$ convox releases promote RTAYZHUSGCJ
Promoting RTAYZHUSGCJ... UPDATING
```

**Comment by beedub** [comment]
@MiguelMoll the description of the cpu feature is incorrect. see https://github.com/convox/rack/pull/779#issuecomment-227304151

**Comment by MiguelMoll** [comment]
Good catch @beedub. I'll update it now.

**Pull request merged**